controllers/atlasmigration/controller.go

```diff
@@ -0,0 +1,298 @@
+)
+
+var (
+	//go:embed *.tmpl
+	tmpls embed.FS
+	tmpl  = template.Must(template.New("atlasmigration").ParseFS(tmpls, "*.tmpl"))
+)
+
+// CLI is the interface used to interact with Atlas CLI
+type CLI interface {
+	Apply(ctx context.Context, data *atlas.ApplyParams) (*atlas.ApplyReport, error)
+}
+
+// Reconciler reconciles a AtlasMigration object
+type Reconciler struct {
+	client.Client
+	CLI    CLI
+	Scheme *runtime.Scheme
+}
+
+// HCLTmplData is the data used to render the HCL template
+// that will be used for Atlas CLI
+type HCLTmplData struct {
+	URL       string
+	Migration struct {
+		Dir string
+	}
+	Cloud struct {
+		URL     string
+		Token   string
+		Project string
+	}
+	Data struct {
+		RemoteDir struct {
+			Name string
+			Tag  string
+		}
+	}
+}
+
+//+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	var (
+		am  dbv1alpha1.AtlasMigration
+		err error
+	)
+
+	// Get AtlasMigration resource
+	if err = r.Get(ctx, req.NamespacedName, &am); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// At the end of reconcile, update the status of the resource base on the error
+	defer func() {
+		if err != nil {
+			log.Error(err, "failed to reconcile")
+		}
+
+		clientErr := r.updateResourceStatus(ctx, am, err)
+		if err != nil {
+			log.Error(clientErr, "failed to update resource status")
+		}
+	}()
+
+	// Only the 'latest' version is supported
+	if am.Spec.Version != "latest" {
+		err = fmt.Errorf("unsupported version: %s", am.Spec.Version)
+		return ctrl.Result{}, nil
+	}
+
+	// Reconcile given resource
+	am.Status, err = r.reconcile(ctx, am)
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// Reconcile the given AtlasMigration resource.
+func (r *Reconciler) reconcile(
+	ctx context.Context,
+	am dbv1alpha1.AtlasMigration) (dbv1alpha1.AtlasMigrationStatus, error) {
+
+	// Build HCL template data
+	hclTmplData, cleanUpDir, err := r.buildHCLTmplData(ctx, am)
+	if err != nil {
+		return dbv1alpha1.AtlasMigrationStatus{}, err
+	}
+	defer cleanUpDir()
+
+	// Create atlas.hcl from template data
+	file, cleanUpFile, err := executeHCLTemplate("migrateconf.tmpl", hclTmplData)
+	if err != nil {
+		return dbv1alpha1.AtlasMigrationStatus{}, err
+	}
+	defer cleanUpFile()
+
+	// Execute Atlas CLI migrate command
+	report, err := r.CLI.Apply(ctx, &atlas.ApplyParams{ConfigURL: file})
+	if err != nil {
+		return dbv1alpha1.AtlasMigrationStatus{}, err
```

Comment:
Need to differentiate between transient and permanent errors, see the `result()` method on the declarative migrations